### PR TITLE
Fuzzy Search

### LINKED ISSUE #266: Add Fuzzy Search
Add a fuzzy search to look up classes. 

Not sure how advanced we want to get with this but bare minimum: user types in department code and number to get a result (e.g. "cs 121" -> COMPSCI 121)

If we collaborate with PeterPortal's fuzzy search, we may be able to do some more advanced stuff. The PeterPortal team currently pays a lot of money per month to use OpenSearch for this. Maybe our backend could use the same api key to fetch that data from the service we're already for. In the long run though, I'd like to build our own search index but this may be quiet hard.
- Nice to have: type in course title and match to a course id (e.g. "Intermediate Programming" -> "ICS 33")
- Super stretch goal: type in concepts from a course description and get course suggestions (e.g. "java" -> "ICS 45J")

### PR DESCRIPTION
**Pull request by ecxyzzy**
## Summary
Implement fuzzy searching with the `websoc-fuzzy-search` module.

## Testing
https://staging-343.antalmanac.com

## To-do
- [x] #345
- [x] Currently, queries manipulate the `RightPane` directly, which means the legacy search form will reflect any autocompleted fields after the user returns from the search view. Not really sure if there's a cleaner way to go about this. (It's fine.)
## Issues
Closes #266.

### CODE REVIEW

File: src/components/SearchForm/FuzzySearch.js [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+import TextField from '@material-ui/core/TextField';
+import Autocomplete from '@material-ui/lab/Autocomplete';
+import search from 'websoc-fuzzy-search';
+import { updateFormValue, resetFormValues } from '../../actions/RightPaneActions';
+
+const emojiMap = {
+    GE_CATEGORY: '🏫', // U+1F3EB :school:
+    DEPARTMENT: '🏢', // U+1F3E2 :office:
+    COURSE: '📚', // U+1F4DA :books:
+    INSTRUCTOR: '🧑', // U+1F9D1 :adult:
```
Comment (ChaseC99) [inline src/components/SearchForm/FuzzySearch.js]: Let's use the teacher emoji
```suggestion
    INSTRUCTOR: '🧑‍🏫',
```

File: src/components/SearchForm/FuzzySearch.js [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+                } else if (emoji === emojiMap.COURSE) {
+                    updateFormValue('deptValue', ident[0].split(' ').slice(0, -1).join(' '));
+                    updateFormValue('deptLabel', ident.join(':'));
+                    updateFormValue('courseNumber', ident[0].split(' ').slice(-1)[0]);
+                } else if (emoji === emojiMap.INSTRUCTOR) {
+                    updateFormValue(
+                        'instructor',
+                        Object.keys(this.state.results).filter((x) => this.state.results[x].name === ident)
+                    );
+                }
```
Comment (ChaseC99) [inline src/components/SearchForm/FuzzySearch.js]: Thoughts on using a switch statement for this?
```suggestion
                switch (emoji) {
                    case emojiMap.GE_CATEGORY:
                        updateFormValue(
                            'ge',
                            `GE-${ident[0].split(' ')[2].replace('(', '').replace(')', '').toUpperCase()}`
                        );
                        break;
                    case emojiMap.DEPARTMENT:
                         updateFormValue('deptValue', ident[0]);
                         updateFormValue('deptLabel', ident.join(':'));
                         break;
                    case emojiMap.COURSE:
                        updateFormValue('deptValue', ident[0].split(' ').slice(0, -1).join(' '));
                        updateFormValue('deptLabel', ident.join(':'));
                        updateFormValue('courseNumber', ident[0].split(' ').slice(-1)[0]);
                        break;
                    case emojiMap.INSTRUCTOR:
                        updateFormValue(
                            'instructor',
                            Object.keys(this.state.results).filter((x) => this.state.results[x].name === ident)
                        );
                        break;
                }
```

**Review by ChaseC99 — CHANGES_REQUESTED** [review]
Overall looking awesome! Focusing on functionality, here are some bugs I found:

- Department field gets updated to the class title when you press the back button on search results 
image

- Multi word parsing still needs a bit of work

| `ics` | `ics `&nbsp; | `ics 3` | `ics 33` |
| --- | --- | --- | --- |
| Looks good, shows ics results. | Once you add a space, nothing gets returned. Should show same results as `ics` | Also shows no results but it should show the "ICS 3: Internet Technologies and their Social Impact" and the 3X series | Looks good |
| image | image | image | image |

File: src/components/SearchForm/FuzzySearch.js (hunk 1 above)
Comment (ecxyzzy) [reply src/components/SearchForm/FuzzySearch.js]: Slight issue with this: the zero-width joiner doesn't actually render on my laptop, so the teacher "emoji" actually displays as two separate emojis, like so:
image

Can you replicate this on your setup?

Comment (ChaseC99) [reply src/components/SearchForm/FuzzySearch.js]: Hmm are you on Windows or Linux? I'm on Mac

Comment (ecxyzzy) [reply src/components/SearchForm/FuzzySearch.js]: I'm on Windows. Did some more experimentation, and it looks like Windows 10's emoji font spec only supports ZWJ on gendered person emojis: 👨‍🏫 (man + school -> male teacher) and 👩‍🏫 (woman + school -> female teacher) render as one emoji, but 🧑‍🏫 (person + school) doesn't.

(For what it's worth, Windows 11 does support ZWJ on [more] gender-neutral person emojis. The downside is having to use Windows 11.)

Comment (devsdevsdevs) [reply src/components/SearchForm/FuzzySearch.js]: Classic windows

**Comment by devsdevsdevs** [comment]
👀

File: src/components/SearchForm/FuzzySearch.js (hunk 1 above)
Comment (ChaseC99) [reply src/components/SearchForm/FuzzySearch.js]: Uh oh so we want to use the male or female one? I feel like the male one looks closer to the gender neutral one

Comment (ecxyzzy) [reply src/components/SearchForm/FuzzySearch.js]: Yeah unfortunately I think it's either choosing between them or using the person emoji 😕

> I feel like the male one looks closer to the gender neutral one

That's what I was thinking too, might be our best choice for now...

Comment (ecxyzzy) [reply src/components/SearchForm/FuzzySearch.js]: After some discussion on Discord Chase suggested to use 🍎 (apple), which sounds like a good enough idea to me, so it has been implemented in 13f7a30.

**Comment by ChaseC99** [comment]
I think this is basically good to go! Let's give it another 24hrs for feedback.

Before merging, we should revert to 3a25e8ff1257148af8099557f603af05ea651e74. 
Even though #348 was merged into main, the "Files changed" tab still shows a lot of diffs and we got a merge conflict after merging `main` into `fuzzy-search`.

**Review by ChaseC99 — APPROVED** [review]
The revert looks good. Let's ship this feature 🎉

**Pull request merged**